README.md
# zcash-android-wallet
An Android wallet using the Zcash Android SDK that is maintained by ECC developers.

### Motivation
[Dogfooding](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) - _transitive verb_ -  is the practice of an organization using its own product. This app was created to help us learn. 

Please take note: the wallet is not an official product by ECC, but rather a tool for learning about our libraries that it is built on. This means that we do not have robust infrasturcture or user support for this application. We open sourced it as a resource to make wallet development easier for the Zcash ecosystem. 

### Setup

To run, clone the repo, open it in Android Studio and press play. It should just work.™

#### Requirements
- [The code](https://github.com/zcash/zcash-android-wallet)
- [Android Studio](https://developer.android.com/studio/index.html) or [adb](https://www.xda-developers.com/what-is-adb/)
- A device or emulator

### Install from Android Studio
1. [Install Android studio](https://developer.android.com/studio/install) and setup an emulator    
    1a. If using a device, be sure to [put it in developer mode](https://developer.android.com/studio/debug/dev-options) to enable side-loading apps
2. `Import` the zcash-android-wallet folder.  
    It will be recognized as an Android project.
3. Press play (once it is done opening and indexing)

### OR Install from the command line
To build from the command line, [setup ADB](https://www.xda-developers.com/install-adb-windows-macos-linux/) and connect your device. Then simply run this and it will both build and install the app:
```bash
cd /path/to/zcash-android-wallet
./gradlew
```

## Disclaimers
There are some known areas for improvement:

- This app is mainly intended for learning and improving the related libraries that it uses. There may be bugs.
- This wallet currently only supports receiving at shielded addresses, which makes it incompatible with wallets that do not support sending to shielded addresses. 
- Traffic analysis, like in other cryptocurrency wallets, can leak some privacy of the user.
- The wallet requires a trust in the server to display accurate transaction information. 
- This app has been developed and run exclusively on `mainnet` it might not work on `testnet`.  

See the [Wallet App Threat Model](https://zcash.readthedocs.io/en/latest/rtd_pages/wallet_threat_model.html)
for more information about the security and privacy limitations of the wallet.

If you'd like to sign up to help us test, reach out on discord and let us know! We're always happy to get feedback!

### License
MIT


CHANGELOG.md
Change Log
==========

Version 1.0.0-alpha74 *(2021-06-21)*
------------------------------------
- New: Added workflow for automatically shielding funds.
- New: Automatically recover from more network failure states.
- New: Link to play store from the build number.
- New: Hide available/total toggle when there are no pending funds.
- New: Easter Egg to access the testnet faucets.
- New: Updated checkpoints for mainnet and testnet.
- Fix: Expand tappable area for showing the balance details.
- Fix: Off by one error when calculating confirmations.
- Fix: Do not show time in transaction details for pending transactions.

Version 1.0.0-alpha72 *(2021-06-07)*
------------------------------------
- New: Address tabs with t-address support [Credit @herou].
- New: Ktlint support [Credit @nighthawk24]
- New: Balance details screen [Credit @herou].
- New: Better balance information around unmined transactions.
- New: Add toggle to show available v. total funds.
- New: Auto-shielding via balance details screen.
- New: 'Ask Later' on feedback popup.
- Fix: Repaired QR scanning on older devices (below API 24).
- Fix: Several of the most frequent crashes reported in bugsnag.
- Fix: Corrected over-sized icon in history.
- Fix: History no longer displays negative balance during initial sync.
- Fix: Errors that prevented sync from working in some situations.
- Fix: Improved support for smaller screens and older devices.

Version 1.0.0-alpha67 *(2021-04-22)*
------------------------------------
- Fix: Crash after entering 24th seed word on certain devices
- Fix: Correct errors on certain devices around biometrics.
- Fix: Display information when invalid addresses are scanned.
- Fix: Prevent memo for transparent transactions [Credit @mandeepbhalothia].
- Fix: Security finding: remove aparent logging [Credit @mandeepbhalothia].
- New: View more info for failed sends [Credit @herou].
- New: Switch away from google libraries for QR code parsing [Credit @herou].
- New: Ability to make testnet releases [Credit @herou].
- New: Updated design for wallet history [Credit @mandeepbhalothia].
- New: Added responsible disclosure document for vulnerabilities [Credit: @zebambam]
- New: Update to latest SDK changes, including using one library instead of two.
- New: Auto-shielding Easter Egg.
- New: Periodically solicit user feedback.
- New: Wallet restore improvements including 'clear' feature.
- New: Rescan/Wipe feature for fixing wallet errors
- New: Add ability to copy the contents of the memo.
- New: Add ability to paste and parse an entire seed phrase at once
- New: Basic support for QRs prefixed with zcash:
- New: Made it easier to add vibration and leveraged it whenever text is copied.
- New: Adds cleanup and removal of failed transactions.
- New: Improved logic for determining the wallet birthday.
- New: Capture performance metrics for scanning.
- New: Additional messaging during a scan for better responsiveness.
- New: Improved handling of critical errors.

Version 1.0.0-alpha43 *(2020-12-20)*
------------------------------------
- Fix: Repaired the upgrade flow, which could not reorg because of missing birthday height
- Fix: Repaired create wallet flow which was being covered by the loading screen

Version 1.0.0-alpha42 *(2020-12-19)*
------------------------------------
- Fix: Correct race condition when launching the app
- Fix: Display loading screen while waiting for app to initialize

Version 1.0.0-alpha41 *(2020-12-19)*
------------------------------------
- New: Upgrade to the latest SDK.
- New: Implements ZIP-313, reducing the default fee from 10,000 to 1,000 zats.
- New: Adds authentication prior to viewing backup seed words.
- New: Adds blockchair as the transaction explorer.
- Fix: Authentication bugs on older devices that were preventing sends and mishandling cancels.
- Fix: Users can now upgrade from seed-only prior versions without crashing or needing to restore.
- Fix: Improved internal metrics for troubleshooting issues.

Version 1.0.0-alpha37 *(2020-10-07)*
------------------------------------
- New: Localization in 5 languages Russian, Italian, Spanish, Chinese and Korean.
- New: Store and sync using just the ViewingKey.
- New: Added QA build flavor for better testing.
- New: Ability to change servers (thanks @Nighthawk!)
- Fix: Critical bug in 3rd-party secure storage library impacting large strings.
- Fix: Devices without PIN can use the wallet again.
- Fix: Developer logs now work on all devices.

Version 1.0.0-alpha34 *(2020-08-28)*
------------------------------------
- New: Implemented transaction detail view.
- New: Updated receive screen and scan screen.
- New: Added optional blockchain explorer with privacy warning.
- Fix: Update key dependencies for performance.
- Fix: Iterated on send flow with lots of improvements and fixes.
- Fix: Trim improperly parsed characters from memos.
- Fix: Keypad stops working when navigating back to home screen.
- Fix: Prevent black screen after failed initialization.

Version 1.0.0-alpha33 *(2020-08-13)*
------------------------------------
- New: Fully removed crashlytics, in favor of bugsnag.
- New: Change the default lightwalletd server.
- New: Switched to the latest SDK.

Version 1.0.0-alpha32 *(2020-08-01)*
------------------------------------
- New: entirely revamped send flow
- New: added biometric authentication support
- New: add robust support for tx cancellation
- New: support precise birthday heights for faster restore
- New: switched to Reply-To standard for memos
- New: improved feedback while scanning QRs
- New: more compatible with memo reply-to formats
- New: update to latest librustzcash crates
- New: checkpoints
- Fix: amount not clearing on return to home screen
- Fix: address cursor resetting while typing
- Fix: app crash when opening application logs
- Fix: limit decimal places to 8 places
- Fix: wallet history now scrolls to the top
- Fix: consistent currency formatting
- Fix: security finding around compromised file system


app/src/main/java/cash/z/ecc/android/feedback/Report.kt
package cash.z.ecc.android.feedback

import cash.z.ecc.android.ZcashWalletApp

object Report {

    object Funnel {
        sealed class Send(stepName: String, step: Int, vararg properties: Pair<String, Any>) : Feedback.Funnel("send", stepName, step, *properties) {
            object AddressPageComplete : Send("addresspagecomplete", 10)
            object MemoPageComplete : Send("memopagecomplete", 20)
            object ConfirmPageComplete : Send("confirmpagecomplete", 30)

            // Beginning of send
            object SendSelected : Send("sendselected", 50)
            object SpendingKeyFound : Send("keyfound", 60)
            object Creating : Send("creating", 70)
            object Cancelled : Send("cancelled", 72)
            class Created(id: Long) : Send("created", 80, "id" to id)
            object Submitted : Send("submitted", 90)
            class Mined(minedHeight: Int) : Send("mined", 100, "minedHeight" to minedHeight)

            // Errors
            abstract class Error(stepName: String, step: Int, val errorCode: Int?, val errorMessage: String?, vararg properties: Pair<String, Any>) : Send("error.$stepName", step, "isError" to true, *properties)
            object ErrorNotFound : Error("notfound", 51, null, "Key not found")
            class ErrorEncoding(errorCode: Int? = null, errorMessage: String? = null) : Error(
                "encode", 71, errorCode, errorMessage,
                "errorCode" to (errorCode ?: -1),
                "errorMessage" to (errorMessage ?: "None")
            )
            class ErrorSubmitting(errorCode: Int? = null, errorMessage: String? = null) : Error(
                "submit", 81, errorCode, errorMessage,
                "errorCode" to (errorCode ?: -1),
                "errorMessage" to (errorMessage ?: "None")
            )
        }

        sealed class Restore(stepName: String, step: Int, vararg properties: Pair<String, Any>) : Feedback.Funnel("restore", stepName, step, *properties) {
            object Initiated : Restore("initiated", 0)
            object SeedWordsStarted : Restore("wordsstarted", 10)
            class SeedWordCount(wordCount: Int) : Restore("wordsmodified", 15, "seedWordCount" to wordCount)
            object SeedWordsCompleted : Restore("wordscompleted", 20)
            object Stay : Restore("stay", 21)
            object Exit : Restore("stay", 22)
            object Done : Restore("doneselected", 30)
            object ImportStarted : Restore("importstarted", 40)
            object ImportCompleted : Restore("importcompleted", 50)
            object Success : Restore("success", 100)
        }

        sealed class UserFeedback(stepName: String, step: Int, vararg properties: Pair<String, Any>) : Feedback.Funnel("feedback", stepName, step, *properties) {
            object Started : UserFeedback("started", 0)
            object Cancelled : UserFeedback("cancelled", 1)
            class Submitted(rating: Int, question1: String, question2: String, question3: String, isSolicited: Boolean) : UserFeedback("submitted", 100, "rating" to rating, "question1" to question1, "question2" to question2, "question3" to question3, "isSolicited" to isSolicited)
        }
    }

    object Error {
        object NonFatal {
            class Reorg(errorBlockHeight: Int, rewindBlockHeight: Int) : Feedback.AppError(
                "reorg",
                "Chain error detected at height $errorBlockHeight, rewinding to $rewindBlockHeight",
                false,
                "errorHeight" to errorBlockHeight,
                "rewindHeight" to rewindBlockHeight
            ) {
                val errorHeight: Int by propertyMap
                val rewindHeight: Int by propertyMap
            }
            class TxUpdateFailed(t: Throwable) : Feedback.AppError("txupdate", t, false)
            abstract class TxError(action: String, val errorCode: Int?, val errorMessage: String?) : Feedback.AppError(
                "tx.$action",
                "Failed to $action transaction due to $errorMessage",
                false,
                "errorCode" to (errorCode ?: 1)
            )
            class TxEncodeError(errorCode: Int?, errorMessage: String?) : TxError("encode", errorCode, errorMessage)
            class TxSubmitError(errorCode: Int?, errorMessage: String?) : TxError("submit", errorCode, errorMessage)
        }
    }

    sealed class Performance(name: String, vararg properties: Pair<String, Any>) : Feedback.MappedAction(
        "metricName" to name,
        "isPerformanceMetric" to true,
        *properties
    ) {
        override val key = "performance.$name"
        override fun toString() = "$key: ${toMap().let { if (it.size > 1) "${it.entries}" else "" }}"

        class ScanRate(network: String, cumulativeItems: Int, cumulativeTime: Long, cumulativeIps: Float) : Performance("scan.bps", "network" to network, "totalBlocks" to cumulativeItems, "totalTime" to cumulativeTime, "blocksPerSecond" to cumulativeIps)
    }

    // placeholder for things that we want to monitor
    sealed class Issue(name: String, vararg properties: Pair<String, Any>) : Feedback.MappedAction(
        "issueName" to name,
        "isIssue" to true,
        *properties
    ) {
        override val key = "issue.$name"
        override fun toString() = "occurrence of ${key.replace('.', ' ')}${toMap().let { if (it.size > 1) " with ${it.entries}" else "" }}"

        // Issues with sending worth monitoring
        object SelfSend : Issue("self.send")
        object TinyAmount : Issue("tiny.amount")
        object MicroAmount : Issue("micro.amount")
        object MinimumAmount : Issue("minimum.amount")
        class TruncatedMemo(memoSize: Int) : Issue("truncated.memo", "memoSize" to memoSize)
        class LargeMemo(memoSize: Int) : Issue("large.memo", "memoSize" to memoSize)
        class MissingViewkey(recovered: Boolean, needle: String, haystack: String, hasKey: Boolean) : Issue(
            "missing.viewkey", "wasAbleToRecover" to recovered, "needle" to needle, "haystack" to haystack, "hasKey" to hasKey
        )
    }

    enum class Screen(val id: String? = null) : Feedback.Action {
        BACKUP,
        HOME,
        HISTORY("wallet.history"),
        TRANSACTION("wallet.transaction"),
        LANDING,
        PROFILE,
        AWESOME,
        FEEDBACK,
        RECEIVE,
        RESTORE,
        SCAN,
        AUTO_SHIELD_FINAL("autoshield.final"),
        AUTO_SHIELD_AVAILABLE("autoshield.available"),
        SEND_ADDRESS("send.address"),
        SEND_CONFIRM("send.confirm"),
        SEND_FINAL("send.final"),
        SEND_MEMO("send.memo");

        override val key = "screen.${id ?: name.toLowerCase()}"
        override fun toString() = "viewed the ${key.substring(7).replace('.', ' ')} screen"
    }

    enum class Tap(val id: String) : Feedback.Action {
        BACKUP_DONE("backup.done"),
        BACKUP_VERIFY("backup.verify"),
        DEVELOPER_WALLET_PROMPT("landing.devwallet.prompt"),
        DEVELOPER_WALLET_IMPORT("landing.devwallet.import"),
        DEVELOPER_WALLET_CANCEL("landing.devwallet.cancel"),
        LANDING_RESTORE("landing.restore"),
        LANDING_NEW("landing.new"),
        LANDING_BACKUP("landing.backup"),
        LANDING_BACKUP_SKIPPED_1("landing.backup.skip.1"),
        LANDING_BACKUP_SKIPPED_2("landing.backup.skip.2"),
        LANDING_BACKUP_SKIPPED_3("landing.backup.skip.3"),
        HOME_PROFILE("home.profile"),
        HOME_HISTORY("home.history"),
        HOME_RECEIVE("home.receive"),
        HOME_BALANCE_DETAIL("home.balance.detail"),
        TAB_LAYOUT("tab.layout"),
        HOME_SCAN("home.scan"),
        HOME_SEND("home.send"),
        HOME_FUND_NOW("home.fund.now"),
        HOME_CLEAR_AMOUNT("home.clear.amount"),
        HISTORY_BACK("history.back"),
        TRANSACTION_BACK("transaction.back"),
        PROFILE_CLOSE("profile.close"),
        AWESOME_OPEN("profile.awesome"),
        AWESOME_CLOSE("awesome.close"),
        AWESOME_SHIELD("awesome.shield"),
        PROFILE_BACKUP("profile.backup"),
        PROFILE_RESCAN("profile.rescan"),
        PROFILE_VIEW_USER_LOGS("profile.view.user.logs"),
        PROFILE_VIEW_DEV_LOGS("profile.view.dev.logs"),
        PROFILE_SEND_FEEDBACK("profile.send.feedback"),
        FEEDBACK_CANCEL("feedback.cancel"),
        FEEDBACK_SUBMIT("feedback.submit"),
        RECEIVE_BACK("receive.back"),
        RESTORE_DONE("restore.done"),
        RESTORE_CLEAR("restore.clear"),
        RESTORE_SUCCESS("restore.success"),
        RESTORE_BACK("restore.back"),
        SCAN_BACK("scan.back"),
        AUTO_SHIELD_FINAL_CLOSE("autoshield.final.close"),
        AUTO_SHIELD_FINAL_DONE("autoshield.final.done"),
        SEND_ADDRESS_MAX("send.address.max"),
        SEND_ADDRESS_NEXT("send.address.next"),
        SEND_ADDRESS_PASTE("send.address.paste"),
        SEND_ADDRESS_REUSE("send.address.reuse"),
        SEND_ADDRESS_BACK("send.address.back"),
        SEND_ADDRESS_DONE_ADDRESS("send.address.done.address"),
        SEND_ADDRESS_DONE_AMOUNT("send.address.done.amount"),
        SEND_ADDRESS_SCAN("send.address.scan"),
        SEND_CONFIRM_BACK("send.confirm.back"),
        SEND_CONFIRM_NEXT("send.confirm.next"),
        SEND_FINAL_EXIT("send.final.exit"),
        SEND_FINAL_RETRY("send.final.retry"),
        SEND_FINAL_CLOSE("send.final.close"),
        SEND_MEMO_INCLUDE("send.memo.include"),
        SEND_MEMO_EXCLUDE("send.memo.exclude"),
        SEND_MEMO_NEXT("send.memo.next"),
        SEND_MEMO_SKIP("send.memo.skip"),
        SEND_MEMO_CLEAR("send.memo.clear"),
        SEND_MEMO_BACK("send.memo.back"),

        SEND_SUBMIT("send.submit"),

        // General events
        COPY_ADDRESS("copy.address"),
        COPY_TRANSPARENT_ADDRESS("copy.address.transparent");

        override val key = "tap.$id"
        override fun toString() = "${key.replace('.', ' ')} button".replace("tap ", "tapped the ")
    }

    enum class NonUserAction(override val key: String, val description: String) : Feedback.Action {
        FEEDBACK_STARTED("action.feedback.start", "feedback started"),
        FEEDBACK_STOPPED("action.feedback.stop", "feedback stopped"),
        SYNC_START("action.feedback.synchronizer.start", "sync started");

        override fun toString(): String = description
    }

    enum class MetricType(override val key: String, val description: String) : Feedback.Action {
        ENTROPY_CREATED("metric.entropy.created", "entropy created"),
        SEED_CREATED("metric.seed.created", "seed created"),
        SEED_IMPORTED("metric.seed.imported", "seed imported"),
        SEED_PHRASE_CREATED("metric.seedphrase.created", "seed phrase created"),
        SEED_PHRASE_LOADED("metric.seedphrase.loaded", "seed phrase loaded"),
        WALLET_CREATED("metric.wallet.created", "wallet created"),
        WALLET_IMPORTED("metric.wallet.imported", "wallet imported"),
        ACCOUNT_CREATED("metric.account.created", "account created"),

        // Transactions
        TRANSACTION_INITIALIZED("metric.tx.initialized", "transaction initialized"),
        TRANSACTION_CREATED("metric.tx.created", "transaction created successfully"),
        TRANSACTION_SUBMITTED("metric.tx.submitted", "transaction submitted successfully"),
        TRANSACTION_MINED("metric.tx.mined", "transaction mined")
    }
}

/**
 * Creates a metric with a start time of ZcashWalletApp.creationTime and an end time of when this
 * instance was created. This can then be passed to [Feedback.report].
 */
class LaunchMetric private constructor(private val metric: Feedback.TimeMetric) :
    Feedback.Metric by metric {
    constructor() : this(
        Feedback
            .TimeMetric(
                "metric.app.launch",
                "app launched",
                mutableListOf(ZcashWalletApp.instance.creationTime)
            )
            .markTime()
    )
    override fun toString(): String = metric.toString()
}

inline fun <T> Feedback.measure(type: Report.MetricType, block: () -> T): T =
    this.measure(type.key, type.description, block)
